### PR TITLE
wifi: Fix enterprise mode

### PR DESCRIPTION
**Pull request by MaochenWang1**
Fix AP cmd can't use by splitting wifi interface into STA and uAP
Fix hostap compile error and support enterprise
Fix wifi sta connection not triggerred issue
Add macro to wrap the interface monitoring code
Fix EAPol packet length changed wrongly

### CODE REVIEW

File: include/zephyr/net/ethernet.h [hunk 1]
```diff
@@ -433,8 +436,11 @@ enum ethernet_if_types {
 	/** IEEE 802.3 Ethernet (default) */
 	L2_ETH_IF_TYPE_ETHERNET,
 
-	/** IEEE 802.11 Wi-Fi*/
-	L2_ETH_IF_TYPE_WIFI,
+	/** IEEE 802.11 Wi-Fi Station */
+	L2_ETH_IF_TYPE_WIFI_STA,
+
+	/** IEEE 802.11 Wi-Fi Soft AP */
+	L2_ETH_IF_TYPE_WIFI_UAP,
```
Comment (jukkar) [inline include/zephyr/net/ethernet.h:443]: This `ethernet_if_type` is used to differentiate Wi-Fi from Ethernet. The Station vs AP selection is a wifi property and this check does not really belong here.

File: modules/hostap/src/supp_api.c [hunk 2]
```diff
@@ -511,8 +511,17 @@ static int wpas_add_and_config_network(struct wpa_supplicant *wpa_s,
 	}
 
 	/* enable and select network */
-	if (!wpa_cli_cmd_v("enable_network %d", resp.network_id)) {
-		goto out;
+	if (mode_ap)
+	{
```
Comment (jukkar) [inline modules/hostap/src/supp_api.c:515]: Please follow zephyr coding style, so the start of the block should be in previous line
```
if (mode_ap) {
....
} else {
....
}
```

File: modules/hostap/Kconfig [hunk 3]
```diff
@@ -175,6 +175,10 @@ config WIFI_NM_WPA_SUPPLICANT_BSS_MAX_IDLE_TIME
 config WIFI_NM_WPA_SUPPLICANT_NO_DEBUG
 	bool "Disable printing of debug messages, saves code size significantly"
 
+config WIFI_NM_WPA_SUPPLICANT_INF_MON
+	bool "Monitor the net mgmt event to add/del interface"
+	default y
```
Comment (jukkar) [inline modules/hostap/Kconfig:180]: Why do we need this, I mean when would we want to turn this interface monitoring off?

File: subsys/net/l2/ethernet/ethernet.c [hunk 4]
```diff
@@ -402,7 +402,10 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 		return net_gptp_recv(iface, pkt);
 	}
 
-	ethernet_update_length(iface, pkt);
+	if (type != NET_ETH_PTYPE_EAPOL)
```
Comment (jukkar) [inline subsys/net/l2/ethernet/ethernet.c:405]: Could you add in the commit
`net: ethernet: Fix EAPol packet length changed wrongly`
a bit more information that answers the question "why" this change is needed.
So something like
```
In ethernet_recv(), for EAPol packet, its length should not be modified
like other IP packets because ....
```
or something like this. Now it is not obvious why this is needed.

**Review by jukkar — COMMENTED** [review]
Looks mostly ok. Thanks especially for the enterprise fixes, clearly this was not tested properly.

File: include/zephyr/net/ethernet.h (hunk 1 above)
Comment (krish2718) [reply include/zephyr/net/ethernet.h:443]: Also, FYI, I have been SAP to represent SoftAP

File: modules/hostap/CMakeLists.txt [hunk 5]
```diff
@@ -372,27 +378,32 @@ zephyr_library_sources_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_CRYPTO_ENTERPRISE
...
-	CONFIG_EAP_LEAP
-	CONFIG_EAP_PSK
-	CONFIG_EAP_FAST
-	CONFIG_EAP_PAX
-	CONFIG_EAP_SAKE
-	CONFIG_EAP_GPSK
-	CONFIG_EAP_PWD
-	CONFIG_EAP_EKE
-	CONFIG_EAP_IKEv2
+	EAP_TLS
```
Comment (krish2718) [inline modules/hostap/CMakeLists.txt:388]: Thanks, this was recently fixed in NCS downstream.

File: modules/hostap/Kconfig (hunk 3 above)
Comment (krish2718) [reply modules/hostap/Kconfig:180]: And without a interface added to the WPA supplicant it's useless.

**Review by krish2718 — CHANGES_REQUESTED** [review]
General comment, multiple changes are clubbed in to a single commit, this is evident by the commit logs having unrelated change descriptions, so, please split them to have a single logical change per-commit.

File: modules/hostap/src/supp_api.c [hunk 6]
```diff
@@ -511,8 +511,17 @@ static int wpas_add_and_config_network(struct wpa_supplicant *wpa_s,
...
+		if (!wpa_cli_cmd_v("enable_network %d", resp.network_id)) {
+			goto out;
+		}
+	}
+	else
+	{
+		if (!wpa_cli_cmd_v("select_network %d", resp.network_id)) {
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:522]: for both AP and STA you can do `enable` and `select` for both.

File: include/zephyr/net/ethernet.h (hunk 1 above)
Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @jukkar, split L2_ETH_IF_TYPE_WIFI into L2_ETH_IF_TYPE_WIFI_STA and L2_ETH_IF_TYPE_WIFI_SAP is a more easy and clean way. Or can I add new "enum wifi_if_types" in wifi.h and add "enum wifi_if_types wifi_if_type" in "struct ethernet_context"? do you have any suggesions?

Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @krish2718 , I will change UAP to SAP, thanks

File: modules/hostap/src/supp_api.c (hunk 2 above)
Comment (MaochenWang1) [reply modules/hostap/src/supp_api.c:515]: will update this

File: modules/hostap/Kconfig (hunk 3 above)
Comment (MaochenWang1) [reply modules/hostap/Kconfig:180]: Hi @jukkar @krish2718 , actually we turn this monitoring off in our chip and currently we don't need to delete interface when the net_if down. 
the setup_interface_monitoring() already call net_if_foreach(iface_cb, ctx), and iface_cb will add_interface.

File: subsys/net/l2/ethernet/ethernet.c (hunk 4 above)
Comment (MaochenWang1) [reply subsys/net/l2/ethernet/ethernet.c:405]: will update

File: modules/hostap/src/supp_api.c (hunk 6 above)
Comment (MaochenWang1) [reply modules/hostap/src/supp_api.c:522]: Hi @krish2718 
the setup_interface_monitoring() already call net_if_foreach(iface_cb, ctx), and iface_cb will add_interface.

Comment (MaochenWang1) [reply modules/hostap/src/supp_api.c:522]: Hi @krish2718 
but select_network is enough for sta to connect, no need the enable_network.
I add the distinguish between sta and AP, because I don't want to change the logic of AP mode, only want to change for sta mode.

File: include/zephyr/net/ethernet.h (hunk 1 above)
Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @krish2718, UAP is the abbreviation of Micro-AP, can we keep this unchanged?

**Comment by MaochenWang1** [comment]
Hi @jukkar @krish2718 
please help review the change and comment replies, thanks

File: modules/hostap/src/supp_main.c [hunk 7]
```diff
@@ -214,8 +215,20 @@ static int add_interface(struct supplicant_context *ctx, struct net_if *iface)
...
-	wpa_s->conf->ap_scan = 1;
+
+	if (eth_ctx->eth_if_type == L2_ETH_IF_TYPE_WIFI_STA) {
+		wpa_s->conf->ap_scan = 1;
+	}
+#ifdef CONFIG_AP
+	else if (eth_ctx->eth_if_type == L2_ETH_IF_TYPE_WIFI_UAP) {
+		wpa_s->conf->ap_scan = 2;
+	}
+#endif
```
Comment (krish2718) [inline modules/hostap/src/supp_main.c:230]: Please move to a separate commit as it's unrelated to this commit log.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: modules/hostap/Kconfig (hunk 3 above)
Comment (MaochenWang1) [reply modules/hostap/Kconfig:180]: Yes, currently we only add interface at the init time and the interface is always present.
add or del interface based on the netif up/down is not used by us, I think add this macro is more clean and user can choose enable it or disable it. Currently disable it is more easy for us to develop upstream supplicant.

File: subsys/net/l2/ethernet/ethernet.c [hunk 8]
```diff
@@ -402,7 +402,9 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 		return net_gptp_recv(iface, pkt);
 	}
 
-	ethernet_update_length(iface, pkt);
+	if (type != NET_ETH_PTYPE_EAPOL) {
```
[the remark this answers is not in the document]
Comment (MaochenWang1) [reply subsys/net/l2/ethernet/ethernet.c:405]: At the beginning, I also want to move the check into ethernet_update_length().
but I can't get the NET_ETH_PTYPE_EAPOL packet type in ethernet_update_length, as the eth header is removed, and the family of eapol packet is also not saved in the pkt, I can't just check the AF_UNSPEC family, right?
so it's more clean and easier to just check here.

File: modules/hostap/src/supp_main.c (hunk 7 above)
Comment (MaochenWang1) [reply modules/hostap/src/supp_main.c:230]: will update, thanks

File: modules/hostap/src/supp_main.c [hunk 9]
```diff
@@ -214,8 +215,20 @@ static int add_interface(struct supplicant_context *ctx, struct net_if *iface)
...
+#ifdef CONFIG_WIFI_NM_WPA_SUPPLICANT_WPS
+	wpa_s->conf->wps_cred_processing = 2;
+#endif
```
[the remark this answers is not in the document]
Comment (MaochenWang1) [reply modules/hostap/src/supp_main.c:233]: will update, thanks

File: include/zephyr/net/ethernet.h (hunk 1 above)
Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @jukkar @krish2718 
about the UAP, I can change it to SAP.
can I add new "enum wifi_if_types"(contains STA and SAP)  in wifi.h? and add "enum wifi_if_types wifi_if_type" in "struct ethernet_context", and keep "enum ethernet_if_types eth_if_type" unchanged?
do you have any suggestions to implement the "STA and SAP"？

Comment (krish2718) [reply include/zephyr/net/ethernet.h:443]: So, far the Wi-Fi management doesn't have any state, it's mostly a relay from applications to the drivers, but now we have a case where the NM (i.e., WPA supplicant) needs to store per-iface data, so, IMHO, instead of putting this in L2-Ethernet context, we can introduce NM private data?

File: modules/hostap/src/supp_api.c (hunk 6 above)
Comment (MaochenWang1) [reply modules/hostap/src/supp_api.c:519]: will update it

File: include/zephyr/net/ethernet.h (hunk 1 above)
Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @krish2718 , I try to introduce NM private data, you mean in this?
image
but I have a problem, in add_interface(), it will call wifi_nm_register_mgd_iface(), if I also add the sta/softap type here, based on the net_if iface, how could I know this interface belongs to sta or softap?

Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h:443]: Hi @krish2718 can I just add enum wifi_if_types"(contains STA and SAP) into struct ethernet_context, and keep enum ethernet_if_types unchanged? it's more easy and clean to implement it.

Comment (krish2718) [reply include/zephyr/net/ethernet.h:443]: > have a case where the NM (i.e., WPA supplicant) needs to store per-iface data

I was proposing per-face data, basically a wrapper around `struct net_if`, something like below (just a proposal).

```
enum iface_type {
          STA = 0,
          SAP,
          P2P_CLI,
          P2P_GO
};
struct nm_mgd_iface {
   unsigned char type; /* bitmap of types, 8 types should suffice or we can make it short */
   struct net_if *iface;
};

struct nm_mgd_iface *mgd_ifaces[];

char nm_wifi_get_type(void *priv);
```

WDYT? Does this solve your problem?

Comment (MaochenWang1) [reply include/zephyr/net/ethernet.h]: Hi @krish2718 
can you review all again, thanks

**Comment by MaochenWang1** [comment]
Hi @krish2718 @jukkar Pls help review again

**Comment by gannicus-gj** [comment]
> Looks mostly ok. Thanks especially for the enterprise fixes, clearly this was not tested properly.

@krish2718 ， so this means that we get your vote, can merge?

File: include/zephyr/net/net_if.h [hunk 10]
```diff
@@ -3013,6 +3013,20 @@ bool net_if_is_wifi(struct net_if *iface);
...
+ * @return Pointer to network interface, NULL if not found.
+ */
+struct net_if *net_if_get_wifi_sta(void);
+
+/**
+ * @brief Get first Wi-Fi network Soft-AP interface.
+ *
+ * @return Pointer to network interface, NULL if not found.
+ */
+struct net_if *net_if_get_wifi_uap(void);
```
Comment (krish2718) [inline include/zephyr/net/net_if.h:3028]:
```suggestion
struct net_if *net_if_get_wifi_sap(void);
```

**Review by krish2718 — COMMENTED** [review]
Mostly LGTM, I would prefer to use the existing term `sap` instead of `uap`, but not too rigid on that.

File: include/zephyr/net/net_if.h (hunk 10 above)
Comment (MaochenWang1) [reply include/zephyr/net/net_if.h]: @krish2718  changed, pls review again, thanks

**Review by krish2718 — DISMISSED** [review]

**Review by jukkar — DISMISSED** [review]

**Comment by krish2718** [comment]
@MaochenWang1 please fix the doc build errors for CI to be green.

**Comment by jukkar** [comment]
This would be useful to have in 3.7 as it fixes compilation issues in certain configurations. It is  also easier for future backports if this is merged to 3.7.

**Comment by jukkar** [comment]
@MaochenWang1 could you check the CI documentation issue so we could try to get this to 3.7

**Comment by MaochenWang1** [comment]
Hi @jukkar @krish2718 , pls review again, I fix the CI documentation issue

**Review by jukkar — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
